File: 1250-longest-common-subsequence/longest-common-subsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function(w1, w2) {
    let n = w1.length, m = w2.length;

    const dp = [...new Array(n+1)].map(a =>[...new Array(m+1)].fill(0))
    for (let arr of dp) arr[0] = 0

    for (let r = 1; r <= n; r++) {
        for (let c = 1; c <= m; c++) {
            let v1 = dp[r-1][c], v2 = dp[r][c-1], v3 = 0
            if (w1[r-1] === w2[c-1]) {
                v3 = dp[r-1][c-1] + 1
            }
            dp[r][c] = Math.max(v1,v2,v3)
        }
    }
    return dp[n][m]
};

File: 1250-longest-common-subsequence/longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length(); int m = text2.length();
        int dp[][] = new int[n+1][m+1];

        for (int i = n-1; i >= 0; i--) {
            for (int j = m-1; j >= 0; j--) {
                int v1 = dp[i+1][j];
                int v2 = dp[i][j+1]; 
                int v3 = 0;
                if (text1.charAt(i) == text2.charAt(j)) {
                    v3 = 1 + dp[i+1][j+1];
                }
                dp[i][j] = Math.max(v1, Math.max( v2,v3));
            }
        }

        return dp[0][0];
    }
}